src/WinRT.Runtime/Marshalers.cs

```diff
@@ -25,55 +26,105 @@ public static void Dispose(this GCHandle handle)
                 handle.Free();
             }
         }
+
+#if !NET
+        public static unsafe ref readonly char GetPinnableReference(this string str)
+        {
+            return ref *(char*)((*(byte**)Unsafe.AsPointer(ref str)) + RuntimeHelpers.OffsetToStringData);
+        }
+#endif
     }
 
-    // TODO: minimize heap allocations for marshalers by eliminating explicit try/finally
-    // and adopting ref structs with non-IDisposable Dispose and 'using var ...' pattern,
-    // as well as passing marshalers to FromAbi by ref so they can be conditionally disposed.
     public class MarshalString
     {
-        private IntPtr _header;
-        public GCHandle _gchandle;
-        public IntPtr _handle;
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct HSTRING_HEADER
+        {
+            IntPtr reserved1;
+            int reserved2;
+            int reserved3;
+            int reserved4;
+            int reserved5;
+        }
+
+        private HSTRING_HEADER _header;
+        private GCHandle _gchandle;
+
+        public struct Pinnable
+        {
+            private HSTRING_HEADER _header;
+            private string _value;
+
+            public Pinnable(string value)
+            {
+                _value = value ?? "";
+                _header = default;
+            }
+
+            public ref readonly char GetPinnableReference()
+            {
+                return ref _value.GetPinnableReference();
+            }
+
+            public unsafe IntPtr GetAbi()
+            {
+                if (_value == "")
+                {
+                    return IntPtr.Zero;
+                }
+                IntPtr hstring;
+                Marshal.ThrowExceptionForHR(Platform.WindowsCreateStringReference(
+                    (char*)Unsafe.AsPointer(ref Unsafe.AsRef(in GetPinnableReference())),
+                    _value.Length,
+                    (IntPtr*)Unsafe.AsPointer(ref _header),
+                    &hstring));
+                return hstring;
+            }
+        }
+
+        public static Pinnable CreatePinnable(string value) => new(value);
+
+        public static IntPtr GetAbi(Pinnable p) => p.GetAbi();

```

Comment:
This is not right. `Pinnable` is passed by value here. The resulting IntPtr is going to be dangling pointer.

Reply:
ah, good catch - I made that change last minute to keep the codegen consistent with object-based marshalers